### PR TITLE
[WIP][Feature] Split dataset for greedynas

### PR DESCRIPTION
**Pull request by humu789**
## Motivation

split dataset for greedynas 

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/datasets/utils.py [hunk 1]
```diff
@@ -1,15 +1,63 @@
+import os
+import random
+
 from torch.utils.data import random_split
 
 
-def split_dataset(dataset):
+def split_dataset_txt(txt_path, save_dir):
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:7]: It is better to make a tool script and put it in the `tools/datasets`

Comment (pppppM) [inline mmrazor/datasets/utils.py:7]: It is better to make a tool script and put it in the `tools/datasets`

File: mmrazor/datasets/utils.py [hunk 2]
```diff
@@ -1,15 +1,63 @@
...
+    index_pool = dict()
+    with open(txt_path, 'r') as f:
+        for line in f:
+            line = line.strip()
+            (filename, index) = line.split(' ')
+            if index not in index_pool:
+                index_pool[index] = []
+            index_pool[index].append(filename)
+
+    assert len(index_pool) == 1000
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:17]: Pls add doc-string to show the `index_pool` example item

Comment (pppppM) [inline mmrazor/datasets/utils.py:17]: Pls add doc-string to show the `index_pool` example item

File: mmrazor/datasets/utils.py [hunk 3]
```diff
@@ -1,15 +1,63 @@
...
+
+    train_txt = list()
+    val_txt = list()
+    for index, filenames in index_pool.items():
+        filenames_val = random.sample(filenames, 50)
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:22]: The `50` seems should be an arg

Comment (pppppM) [inline mmrazor/datasets/utils.py:22]: The `50` seems should be an arg

File: mmrazor/datasets/utils.py [hunk 4]
```diff
@@ -1,15 +1,63 @@
+import os
+import random
+
 from torch.utils.data import random_split
 
 
-def split_dataset(dataset):
+def split_dataset_txt(txt_path, save_dir):
+    index_pool = dict()
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:8]: `index_pool` -> `class2images` ?

Comment (pppppM) [inline mmrazor/datasets/utils.py:8]: `index_pool` -> `class2images` ?

File: mmrazor/datasets/utils.py [hunk 5]
```diff
@@ -1,15 +1,63 @@
...
+            line = line.strip()
+            (filename, index) = line.split(' ')
+            if index not in index_pool:
+                index_pool[index] = []
+            index_pool[index].append(filename)
+
+    assert len(index_pool) == 1000
+
+    train_txt = list()
+    val_txt = list()
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:20]: The `train/val_txt` should be `part_xx`.
This script can also be reused when supporting semi-supervised algorithms in the future.

Comment (pppppM) [inline mmrazor/datasets/utils.py:20]: The `train/val_txt` should be `part_xx`.
This script can also be reused when supporting semi-supervised algorithms in the future.

File: mmrazor/datasets/utils.py [hunk 6]
```diff
@@ -1,15 +1,63 @@
...
+    for index, filenames in index_pool.items():
+        filenames_val = random.sample(filenames, 50)
+        for name in filenames_val:
+            filenames.remove(name)
+        filenames_train = filenames
+        for name in filenames_train:
+            train_txt.append(f'{name} {index}')
+        for name in filenames_val:
+            val_txt.append(f'{name} {index}')
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:29]: The class-balanced sample should be configurable.
Some algorithms don't split the dataset in this way.

Comment (pppppM) [inline mmrazor/datasets/utils.py:29]: The class-balanced sample should be configurable.
Some algorithms don't split the dataset in this way.

File: mmrazor/datasets/utils.py [hunk 7]
```diff
@@ -1,15 +1,63 @@
...
+
+    assert len(val_txt) == 50000
+
+    with open(os.path.join(save_dir, 'greedynas_train.txt'), 'w') as f:
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:33]: `greedynas_train.txt` should be `train/val_part_1.txt` according to the  origin txt name

Comment (pppppM) [inline mmrazor/datasets/utils.py:33]: `greedynas_train.txt` should be `train/val_part_1.txt` according to the  origin txt name

File: mmrazor/datasets/utils.py [hunk 8]
```diff
@@ -1,15 +1,63 @@
...
+        for line in train_txt:
+            f.writelines(f'{line}\n')
+    with open(os.path.join(save_dir, 'greedynas_val.txt'), 'w') as f:
```
Comment (pppppM) [inline mmrazor/datasets/utils.py:36]: `greedynas_val.txt` should be `train/val_part_2.txt` according to the  origin txt name

Comment (pppppM) [inline mmrazor/datasets/utils.py:36]: `greedynas_train.txt` should be `train/val_part_2.txt` according to the  origin txt name

**Review by pppppM — CHANGES_REQUESTED** [review]

File: mmrazor/datasets/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (humu789) [reply mmrazor/datasets/utils.py]: `split_imagenet_dataset` is better

**Pull request closed without merging**